### PR TITLE
added code for Coffee Addicts test

### PR DESCRIPTION
**Pull request by alexandravirvara**

### CODE REVIEW

File: src/model/Coordinate.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+     *
+     **** arg: the String to be converted to a Double
+     **** returns the Double corresponding to the given coordinate as a String, if the coordinate is valid,
+     *              or an invalid coordinate otherwise
+     */
+    private fun convertToDouble(arg: String): Double {
+        return try {
+            arg.toDouble()
+        } catch (nfe: NumberFormatException) {
+            maxLongitude + 1
```
Comment (alexandru-calinoiu) [inline src/model/Coordinate.kt:62]: Not sure what this is used for?

File: src/model/FileParser.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
+import java.io.File
+
+/**
+ * Defines a FileParser object, containing different operations on the content of a given .csv file
+ *
+ * created by AlexandraV on 12/25/2018
+ */
+data class FileParser(var fileName: String, var userCoordinate: Coordinate) {
```
Comment (alexandru-calinoiu) [inline src/model/FileParser.kt:8]: I don't believe this file is used anymore. You should remove it as it fails to compile now..

File: src/parser/FileParser.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import Coordinate
+import java.io.File
+
+/**
+ * Defines a parser.FileParser object, containing different operations on the content of a given .csv file
+ *
+ * created by AlexandraV on 12/25/2018
+ */
+data class FileParser(var fileName: String, var userCoordinate: Coordinate) {
+    val coffeeShopsList: MutableList<CoffeeShop> = mutableListOf()
```
Comment (alexandru-calinoiu) [inline src/parser/FileParser.kt:13]: Most probably this should be private also, any reason is not?

File: src/parser/FileParser.kt [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    /**
+     * inserts a CoffeeShop object into the list. The first ${n} elements of the list must be the first ${n} closest shops to the user
+     * and they must be ordered by distance, from the closest to the furthest
+     *
+     * *** coffeeShopName: the name of the coffeeShop
+     * *** coffeeShopCoordinateX: the X coordinate of the coffeeShop
+     * *** coffeeShopCoordinateY: the Y coordinate of the coffeeShop
+     * *** n : the number of elements from the beginning of the list, ordered by their distance to the user
+     */
+    fun insertShopIntoList(coffeeShopName: String, coffeeShopCoordinateX: String, coffeeShopCoordinateY: String, n: Int) {
```
Comment (pirvudoru) [inline src/parser/FileParser.kt:67]: Why should this method be public?

File: src/model/FileParser.kt [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+     */
+    fun parseCoffeeShopsList(n: Int) {
+        val invalidCoffeeShopsList: MutableList<String> = mutableListOf()
+        if (lines != null) {
+            var lineNumber = 1
+            lines.forEach { it ->
+                val result: List<String> = it.split(",").map { it.trim() }
+                if (isCSVLineValid(result)) {
+                    val shopCoordinate = Coordinate(result[1], result[2])
+                    insertShopIntoList(CoffeeShop(result[0], shopCoordinate.distanceTo(userCoordinate.xCoordinate, userCoordinate.yCoordinate)), n)
```
Comment (pirvudoru) [inline src/model/FileParser.kt:24]: Why prefer passing in x and y to distanceTo instead of a coordinate object?

File: src/parser/FileParser.kt [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        if (n > coffeeShopsList.size && !coffeeShopInserted) {
+            coffeeShopsList.add(coffeeShop)
+        }
+    }
+
+    /**
+     * Checks if a line is valid, of the form: CoffeeShopName, X ShopCoordinate, Y ShopCoordinate
+     */
+    fun isCSVLineValid(line: List<String>): Boolean {
+        if (line.size != 3) {
```
Comment (alexandru-calinoiu) [inline src/parser/FileParser.kt:93]: This code is very hard to followup, [multi returns](https://www.yegor256.com/2015/08/18/multiple-return-statements-in-oop.html) statements are considered a bad practice, can you think of a way to avoid them?

File: src/parser/FileParser.kt (hunk 4 above)
Comment (alexandru-calinoiu) [inline src/parser/FileParser.kt:67]: A couple of comments about this methods signature:

- it does not have an access modifier, most likely should be private
- it receives x, and y and strings although there is a model to hold Coordinates, why is that?
- the `n` parameter not very expressive, I think it means the number of closes shops, the naming can be improved.

File: test/model/CoordinateTest.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+fun isCoordinateValid_Valid() {
+    val coordinate1 = Coordinate("45.0", "30.02")
+    val coordinate2 = Coordinate("0.0", "0.3")
+    val coordinate3 = Coordinate("-90.0", "180.0")
+    val coordinate4 = Coordinate("90.0", "-180.0")
+
+    assert(coordinate1.isCoordinateValid())
+    assert(coordinate2.isCoordinateValid())
+    assert(coordinate3.isCoordinateValid())
+    assert(coordinate4.isCoordinateValid())
```
Comment (pirvudoru) [inline test/model/CoordinateTest.kt:35]: Sometimes it's ok to have multiple asserts per test, but in this case, the asserts/acts are independent so it's preferable to separate them out.

File: test/model/CoordinateTest.kt [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import org.testng.annotations.Test
+import kotlin.test.assertEquals
+
+/**
+ * created by AlexandraV on 12/25/2018
+ */
+@Test
+fun isCoordinateValid_bothInvalid() {
+    val coordinate = Coordinate("877.0", "956.256")
+    assert(!coordinate.isCoordinateValid()) { "x in [-90, 90], y in [-180, 180]" }
```
Comment (pirvudoru) [inline test/model/CoordinateTest.kt:10]: Why would the constructor allow you to create the object if it the values are invalid?

File: src/parser/FileParser.kt [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+     *
+     * *** coffeeShopName: the name of the coffeeShop
+     * *** coffeeShopCoordinateX: the X coordinate of the coffeeShop
+     * *** coffeeShopCoordinateY: the Y coordinate of the coffeeShop
+     * *** n : the number of elements from the beginning of the list, ordered by their distance to the user
+     */
+    fun insertShopIntoList(coffeeShopName: String, coffeeShopCoordinateX: String, coffeeShopCoordinateY: String, n: Int) {
+        val coffeeShop = CoffeeShop(coffeeShopName, Coordinate(coffeeShopCoordinateX, coffeeShopCoordinateY), Coordinate(userCoordinate.x, userCoordinate.y))
+        var coffeeShopInserted = false
+        // check if the current coffee shop enters the "top n" coffee shops according to their distance and insert it on the right position
```
Comment (alexandru-calinoiu) [inline src/parser/FileParser.kt:70]: Code comments are considered a [code smell](https://blog.codinghorror.com/code-smells/). The best practice is to focus on the `why` not on the `what`, but better yet the code should be refactored to not need comments.

Do you really use this amount of comments in your production code?

File: test/model/CoordinateTest.kt [hunk 10]
```diff
@@ -0,0 +1,42 @@
+import org.testng.annotations.Test
+import kotlin.test.assertEquals
+
+/**
+ * created by AlexandraV on 12/25/2018
+ */
+@Test
+fun isCoordinateValid_bothInvalid() {
+    val coordinate = Coordinate("877.0", "956.256")
```
Comment (pirvudoru) [inline test/model/CoordinateTest.kt:9]: What are the pros/cons of having `string` vs `double` passed into the constructor of "Coordinate" ?

File: src/model/CoffeeShop.kt [hunk 11]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Defines a CoffeeShop object, containing a CoffeeShop name, its coordinates and a given user's coordinates
+ *
+ * created by AlexandraV on 12/26/2018
+ */
+data class CoffeeShop(val shopName: String, val shopCoordinate: Coordinate, val userCoordinate: Coordinate) : Comparable<CoffeeShop> {
```
Comment (pirvudoru) [inline src/model/CoffeeShop.kt:6]: What if you want to compare the same coffee shop to more than one user? How would the CoffeeShop class look like then?

File: src/CoffeeAddictsMain.kt [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+ */
+fun main(args: Array<String>) {
+    if (!areCommandLineArgumentsValid(args)) {
+        return
+    } else {
+        // read the coordinate of the user
+        val userCoordinate = Coordinate(args[0], args[1])
+
+        // read content of .csv file
+        val fileParser = FileParser(args[2], userCoordinate)
```
Comment (pirvudoru) [inline src/CoffeeAddictsMain.kt:15]: What is the relation between a "FileParser" and a `userCoordinate`?

File: src/CoffeeAddictsMain.kt [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        fileParser.parseCoffeeShopsList(3)
```
Comment (pirvudoru) [inline src/CoffeeAddictsMain.kt:16]: By the looks of the usage of this method, it as side effects and by passing in `3` as argument looks like does more things that it should. One would expect that `parseCoffeeShops` would process some kind of serialized data and return a list of coffee shops objects.

File: src/parser/FileParser.kt [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+            } else {
+                if (coffeeShopsList.size > 0) {
+                    coffeeShopsList.forEach { println("${it.shopName}, ${it.shopDistanceToUser}") }
+                    println("\nThe number of CoffeeShops you requested ($n) exceeds the number of valid shops in the .csv file. Only ${coffeeShopsList.size} result(s) printed")
+                } else {
+                    println("There are no valid coffee shops in $fileName to be displayed")
+                }
+            }
+        }
+        // write all invalid lines (if there are any) in a file
```
Comment (pirvudoru) [inline src/parser/FileParser.kt:47]: Re comments:

https://refactoring.guru/smells/comments
https://blog.submain.com/field-guide-code-comments/

File: src/parser/FileParser.kt [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        if (!invalidCoffeeShopsList.isEmpty()) {
+            println("\nAll invalid lines found in $fileName are written in invalid_coffee_shops.txt")
+            File("invalid_coffee_shops.txt").bufferedWriter().use { out ->
+                invalidCoffeeShopsList.forEach {
+                    out.write(it)
+                }
+            }
+        }
```
Comment (pirvudoru) [inline src/parser/FileParser.kt:55]: This method is quite hard to follow and does more things that it should. One shouldn't need comments to explain what the code does.

File: src/model/FileParser.kt [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+                } else {
+                    invalidCoffeeShopsList.add("Line $lineNumber: $it \n")
+                }
+                lineNumber++
+            }
+
+            if (n <= coffeeShopsList.size) {
+                coffeeShopsList.subList(0, n).forEach {
+                    println("${it.shopName}, ${it.shopDistanceToUser}")
```
Comment (pirvudoru) [inline src/model/FileParser.kt:33]: What is the problem you are trying to solve by having `println`s all over the place?

**Review by pirvudoru — CHANGES_REQUESTED** [review]
Thanks for submitting your solution. 

Please check comments and the following links:

What is TDD: http://agiledata.org/essays/tdd.html
How to write great Git commits: https://chris.beams.io/posts/git-commit/

File: src/parser/FileParser.kt [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+     * *** coffeeShopCoordinateX: the X coordinate of the coffeeShop
+     * *** coffeeShopCoordinateY: the Y coordinate of the coffeeShop
+     * *** n : the number of elements from the beginning of the list, ordered by their distance to the user
+     */
+    fun insertShopIntoList(coffeeShopName: String, coffeeShopCoordinateX: String, coffeeShopCoordinateY: String, n: Int) {
+        val coffeeShop = CoffeeShop(coffeeShopName, Coordinate(coffeeShopCoordinateX, coffeeShopCoordinateY), Coordinate(userCoordinate.x, userCoordinate.y))
+        var coffeeShopInserted = false
+        // check if the current coffee shop enters the "top n" coffee shops according to their distance and insert it on the right position
+        for (i in 0..(minOf(n, coffeeShopsList.size) - 1)) {
+            if (coffeeShop < coffeeShopsList[i]) {
```
Comment (alexandru-calinoiu) [inline src/parser/FileParser.kt:72]: This is a interesting approach.

Unfortunately for me it does not make any sens to compare coffee shops, because actually they get compared based on the distance from the current user.

There a couple of things that I think can be improved:
- coffe shop object should not hold the coordonates of the user, it's not its concern
- coffe shop's can be comparable but not to get the distance, I can imagine a dsl like `userCoordonates.distanceTo(coffeShoopCoordonates)`. The distance calculation should be the concern of some other class.

File: src/parser/FileParser.kt [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+    private val lines: List<String>? = File(fileName).useLines { it.toList() } // the reader is closed automatically
+
+    /**
+     *  parses the lines of the .csv file and converts them into CoffeeShops objects, if possible
+     *  All the lines that cannot be converted to CoffeeShops objects will be written into  invalid_coffee_shops.txt file
+     */
+    fun parseCoffeeShopsList(n: Int) {
+        val invalidCoffeeShopsList: MutableList<String> = mutableListOf()
+        if (lines != null) {
+            var lineNumber = 1
```
Comment (alexandru-calinoiu) [inline src/parser/FileParser.kt:23]: No point on holding the `lineNumber` one can use [forEachIndexed()](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/for-each-indexed.html)

File: src/parser/FileParser.kt [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+                val result: List<String> = it.split(",").map { it.trim() }
+                if (isCSVLineValid(result)) {
+                    insertShopIntoList(result[0], result[1], result[2], n)
+                } else {
+                    if (!it.trim().isEmpty()) invalidCoffeeShopsList.add("Line $lineNumber: $it \n")
+                }
+                lineNumber++
+            }
+
+            if (n <= coffeeShopsList.size) {
```
Comment (alexandru-calinoiu) [inline src/parser/FileParser.kt:34]: I don't think the list size can be larger that n, as the insertShopIntoList has `minOf(n, coffeeShopsList.size) - 1`

File: src/parser/FileParser.kt [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+import java.io.File
+
+/**
+ * Defines a parser.FileParser object, containing different operations on the content of a given .csv file
+ *
+ * created by AlexandraV on 12/25/2018
+ */
+data class FileParser(var fileName: String, var userCoordinate: Coordinate) {
+    val coffeeShopsList: MutableList<CoffeeShop> = mutableListOf()
+    private val lines: List<String>? = File(fileName).useLines { it.toList() } // the reader is closed automatically
```
Comment (alexandru-calinoiu) [inline src/parser/FileParser.kt:14]: What about using a delegate here like [lazy](https://kotlinlang.org/docs/reference/delegated-properties.html#lazy) do you think that will be appropriate?

File: src/parser/FileParser.kt [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+
+    /**
+     *  parses the lines of the .csv file and converts them into CoffeeShops objects, if possible
+     *  All the lines that cannot be converted to CoffeeShops objects will be written into  invalid_coffee_shops.txt file
+     */
+    fun parseCoffeeShopsList(n: Int) {
```
Comment (alexandru-calinoiu) [inline src/parser/FileParser.kt:20]: This method dose so much more that the name suggests:
- it will parse file
- validate file
- calculate results
- print the results and errors

Let's talk about how to better express this flow in code, do you have some thoughts?

**Review by alexandru-calinoiu — COMMENTED** [review]
Good effort, I like the fact that the code is broke down into multiple classes and it tries to separate concerns.

The big miss is that the test driven requirement from the email was totally ignored, there is no evidence that the red-green-refactor cycle was used.

**Comment by alexandru-calinoiu** [comment]
Followeup via email.
